Data/Graph/Inductive/Query/SP/DAG.hs

```diff
@@ -0,0 +1,200 @@
+    , longestPathBetweenNodes
+    , longestPathFromSource
+    , longestPath
+    , tracePath
+    , runPaths
+    , Number
+) where
+
+import Prelude hiding (read)
+import Data.Graph.Inductive.Graph
+import Data.Vector (Vector,maxIndex,freeze,(!))
+import Data.Vector.Mutable (IOVector,write,set,read,new)
+import Control.Monad (MonadPlus,mplus,mzero)
+import Data.Foldable (foldl',Foldable)
+import Data.Functor ((<$>))
+-- | A wrapper for any number, including Infinity and NegativeInfinity
+--   options.
+data Number a = Infinity | NegativeInfinity | Number a deriving (Show, Eq)
+
+instance Ord a => Ord (Number a) where
+    compare Infinity Infinity = EQ
+    compare Infinity _ = GT
+    compare _ Infinity = LT
+    compare NegativeInfinity NegativeInfinity = EQ
+    compare NegativeInfinity _ = LT
+    compare _ NegativeInfinity = GT
+    compare (Number x) (Number y) = compare x y
+
+instance Functor Number where
+    fmap f (Number x) = Number $ f x
+    fmap _ Infinity = Infinity
+    fmap _ NegativeInfinity = NegativeInfinity
+
+-- | Returns the shortest path between a start node and an end node.
+shortestPathBetweenNodes :: (Graph gr, Real b, Foldable f)
+    => Node -- ^ Start
+    -> Node -- ^ Destination
+    -> Int  -- ^ Length of topsort
+    -> f Node -- ^ Topsort
+    -> gr a b -- ^ Graph
+    -> IO Path
+shortestPathBetweenNodes s d l t g = do
+    (_,predecessor) <- runPaths (Just s) (Number 0) Infinity (>) l t g
+    return $ tracePath d predecessor
+
+-- | Returns the longest path between a start node and an end node.
+longestPathBetweenNodes :: (Graph gr, Real b, Foldable f)
+    => Node -- ^ Start
+    -> Node -- ^ Destination
+    -> Int  -- ^ length of topsort
+    -> f Node -- ^ topsort
+    -> gr a b -- ^ Graph
+    -> IO Path
+longestPathBetweenNodes s d l t g = do
+    (_,predecessor) <- runPaths (Just s) (Number 0) NegativeInfinity (<) l t g
+    return $ tracePath d predecessor
+
+-- | Returns the longest possible path from a source node.
+longestPathFromSource :: (Graph gr, Real b, Foldable f)
+    => Node -- ^ Source node
+    -> Int  -- ^ Length of topsort
+    -> f Node -- ^ Topsort
+    -> gr a b -- ^ Graph
+    -> IO Path
+longestPathFromSource s l t g = do
+    (dist,predecessor) <- runPaths (Just s) (Number 0) NegativeInfinity (<) l t g
+    return $ tracePath (maxIndex dist) predecessor
+
+-- | Returns the longest possible path in the graph.
+longestPath :: (Graph gr, Real b, Foldable f)
+    => Int  -- ^ length of topsort
+    -> f Node -- ^ topsort
+    -> gr a b
+    -> IO Path
+longestPath l t g = do
+    (dist,predecessor) <- runPaths Nothing (Number 0) (Number 0) (<) l t g
+    return $ tracePath (maxIndex dist) predecessor
+
+-- | Traces a path backwards through a vector of predecessors,
+--   where a vector of predecessors is a vector such that the value 
+--   in the vector at any position n where n is a node in the graph 
+--   is the predecessor to that node.
+tracePath :: MonadPlus m 
```

Comment:
1) Put constraint in parentheses

2) Why are you returning an arbitrary `MonadPlus` rather than just a `Path` value?  Is it likely that someone will actually want a different `MonadPlus` instance?

Note that if you just use `Path`, then:

1) This function becomes pure
2) The code is simpler: line 110 becomes just `n : tracePath node v` (and 112 is just `[n]`).

(Not to mention that the comment `MonadPlus of Path` doesn't really make sense to me.)